File: deploy-typesense.py
import typesense
import os
from pathlib import Path
import json
import binascii


def get_field_defaults_for_schema_file(schema_file):
    return [(prop, [] if details['type'] == 'array' else '') for prop, details in json.load(open(schema_file, encoding = 'utf-8'))['properties'].items()]

def process_record(path, fields):
    record = json.load(path.open(encoding = 'utf-8'))
    record = {prop: (str(value) if type(value) is bool else value) for prop, value in record.items()}
    record['id'] = record['slug']
        
    # Typesense *expects* an int (or float) as a sortable field. We don't have that, so we use this rather ugly hack for now.
    record['sort-index'] = int(binascii.hexlify(str.encode(record['id']))[:7], 16)

    # Typesense also expects all fields declared in the schema to have a value.
    for field, default in fields:
        if field not in record: record[field] = default
    
    return record

def document_update(index, record):
    # TODO: Let's hope Typesense implements record updates in the future.
    try:
        client.collections[index].documents[record['id']].delete()
    except typesense.exceptions.ObjectNotFound:
        pass
    client.collections[index].documents.create(record)

def deploy_index(index, directory, fields):
    entrylist = [json.loads(entry) for entry in client.collections[index].documents.export() if len(entry) > 0]
    entries = {entry['id']: entry for entry in entrylist}
    
    pathlist = Path(directory).glob('**/*.json')
    for path in pathlist:
        record = process_record(path, fields)
        
        if not record['id'] in entries:
            print('Creating ' + record['id'])
            client.collections[index].documents.create(record)
        else:
            if record != entries[record['id']]:
                print('Updating ' + record['id'])
                document_update(index, record)
            del entries[record['id']]
    
    for id in entries.keys():
        print('Deleting ' + id)
        try:
            client.collections[index].documents[id].delete()
        except typesense.exceptions.ObjectNotFound:
            pass

client = typesense.Client({
    'master_node': {
        'host': 'localhost',
        'port': '8108',
        'protocol': 'http',
        'api_key': os.environ['TYPESENSE_API_KEY']
    },
    'timeout_seconds': 2
})

companies_fields = get_field_defaults_for_schema_file('schema.json')
svas_fields = get_field_defaults_for_schema_file('schema-supervisory-authorities.json')

deploy_index('companies', 'companies', companies_fields)
deploy_index('supervisory-authorities', 'supervisory-authorities', svas_fields)

File: deploy-typesense.js
const Typesense = require('typesense');
const chunk_array = require('chunk');
const bent = require('bent');
const glob = require('glob');
const path = require('path');

const client = new Typesense.Client({
    nodes: [
        {
            host: 'search-backend.datenanfragen.de',
            port: '443',
            protocol: 'https',
        },
    ],

    apiKey: process.env.TYPESENSE_API_KEY,
    connectionTimeoutSeconds: 10,
});
const post = bent('POST', 'json', 200, 'https://search-backend.datenanfragen.de', {
    'X-TYPESENSE-API-KEY': process.env.TYPESENSE_API_KEY,
});

const getFieldsForSchema = function (schema) {
    let fields = [{ name: 'sort-index', type: 'int32' }];

    for (const [field, details] of Object.entries(schema.properties)) {
        let prop_type = '';
        if (details.type === 'array') {
            if (details.items && details.items.type !== 'string') continue;
            else prop_type = 'string[]';
        } else if (details.type === 'boolean') prop_type = 'bool';
        else prop_type = 'string';

        fields.push({ name: field, type: prop_type, optional: !schema.required.includes(field) });
    }

    return fields;
};

// TODO: No. Please. We *need* to find something better than this. :'(
const intifySlug = (slug) =>
    Math.min(
        Number(
            Array.from(new TextEncoder().encode(slug))
                .slice(0, 7)
                .map((i) => i.toString().padStart(3, '0'))
                .join('')
        ),
        2147483640
    );

const setupCollection = async function (collection_name, schema_filename) {
    const schema = require(path.resolve(schema_filename));
    const fields = getFieldsForSchema(schema);

    await client.collections().create({
        name: collection_name,
        fields,
        default_sorting_field: 'sort-index',
    });
};

const deploy = async function (index, schema_filename, directory) {
    // The main collections are only aliases to the actual collection (named with a Unix timestamp).
    const collection_name = `${index}_${Date.now()}`;
    await setupCollection(collection_name, schema_filename);

    // Grab records and prepare for Typesense.
    const records = glob
        .sync(`./${directory}/*.json`)
        .map((file) => require(path.resolve(file)))
        .map((obj) => ({ ...obj, 'sort-index': intifySlug(obj.slug) }))
        .map((obj) => JSON.stringify(obj));

    // Insert in chunks of 1000 records (Typesense recommends packages < 1MB).
    for (const chunk of chunk_array(records, 1000)) {
        // TODO: Let's hope the Typesense JS library will implement this at some point.
        const result = await post(`/collections/${collection_name}/documents/import`, chunk.join('\n'));

        if (!result.success || result.num_imported !== chunk.length) {
            console.error(result);
            throw 'Importing failed!';
        }
    }

    // Update the alias to the new collection only after importing successfully.
    await client.aliases().upsert(index, { collection_name });
    console.log('Collection', index, 'is now aliased to', collection_name);

    // Garbage-collect the old collections, keeping the last three just in case.
    const collections_to_delete = (await client.collections().retrieve())
        .map((c) => c.name)
        .filter((c) => c.startsWith(index + '_'))
        .filter((c) => c !== collection_name)
        .sort()
        .slice(0, -3);

    for (const c of collections_to_delete) {
        console.log('Deleting old collection', c);
        await client.collections(c).delete();
    }
};

const main = async function () {
    await deploy('companies', 'schema.json', 'companies');
    await deploy('supervisory-authorities', 'schema-supervisory-authorities.json', 'supervisory_authorities');
};

main();
